src/Validation.hs

```diff
@@ -0,0 +1,395 @@
+{- |
+Copyright:  (c) 2014 Chris Allen, Edward Kmett
+            (c) 2018-2020 Kowainik
+SPDX-License-Identifier: MPL-2.0
+Maintainer: Kowainik <xrom.xkov@gmail.com>
+
+Lightweight pure validation based on 'Applicative' and 'Selective' functors.
+
+'Validation' is a monoidal sibling to 'Either' but 'Validation' doesn't have a
+'Monad' instance. 'Validation' allows to accumulate all errors instead of
+short-circuting on the first error so you can display all possible errors at
+once.
+
+Common use-cases include:
+
+1. Validating each input of a form with multiple inputs.
+2. Performing multiple validations of a single value.
+
+Instances of different standard typeclasses provide various semantics:
+
+1. 'Functor': change the type inside 'Success'.
+2. 'Bifunctor': change both 'Failure' and 'Success'.
+3. 'Applicative': apply function to values inside 'Success' and accumulate
+   errors inside 'Failure'.
+4. 'Semigroup': accumulate both 'Failure' and 'Success' with '<>'.
+5. 'Monoid': 'Success' that shores 'mempty'.
+6. 'Alternative': return first 'Success' or accumulate all errors inside 'Failure'.
+-}
+
+{-# LANGUAGE CPP                  #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Validation
+       ( -- * How to use
```

Comment:
I would suggest moving this section right after the data type introduction, so first, we define all then explain how and where it could be used. What do you think?

Reply:
Great idea!